From 5761f02ec9334d334e1ed29294c018b3f0987f63 Mon Sep 17 00:00:00 2001
From: Iain Lane <iain.lane@canonical.com>
Date: Mon, 24 Nov 2014 11:31:08 +0000
Subject: [PATCH 2/2] Only use a header bar in GNOME (shell)

Otherwise use a normal titlebar and use the headerbar like a toolbar.

https://bugzilla.gnome.org/show_bug.cgi?id=740617
---
 src/nautilus-bookmarks-window.c | 15 ++++++++++-----
 src/nautilus-window.c           | 13 ++++++++++++-
 2 files changed, 22 insertions(+), 6 deletions(-)

diff --git a/src/nautilus-bookmarks-window.c b/src/nautilus-bookmarks-window.c
index ad4393b..889ff9c 100644
--- a/src/nautilus-bookmarks-window.c
+++ b/src/nautilus-bookmarks-window.c
@@ -689,6 +689,7 @@ nautilus_bookmarks_window_constructed (GObject *object)
 	GtkWidget *headerbar;
 	GtkTreeViewColumn *col;
 	GtkCellRenderer *rend;
+	const gchar *desktop;
 
 	G_OBJECT_CLASS (nautilus_bookmarks_window_parent_class)->constructed (object);
 
@@ -703,13 +704,17 @@ nautilus_bookmarks_window_constructed (GObject *object)
 		return;
 	}
 
-	headerbar = gtk_header_bar_new ();
-	gtk_header_bar_set_show_close_button (GTK_HEADER_BAR (headerbar), TRUE);
-	gtk_widget_show (headerbar);
-
 	window = GTK_WINDOW (object);
+
+	desktop = g_getenv ("XDG_CURRENT_DESKTOP");
+	if (desktop && strstr (desktop, "GNOME")) {
+		headerbar = gtk_header_bar_new ();
+		gtk_header_bar_set_show_close_button (GTK_HEADER_BAR (headerbar), TRUE);
+		gtk_widget_show (headerbar);
+		gtk_window_set_titlebar (window,  headerbar);
+	}
+
 	gtk_window_set_title (window, _("Bookmarks"));
-	gtk_window_set_titlebar (window,  headerbar);
 	gtk_window_set_type_hint (window, GDK_WINDOW_TYPE_HINT_DIALOG);
 	gtk_window_set_default_size (window, 
 				     BOOKMARKS_WINDOW_INITIAL_WIDTH, 
diff --git a/src/nautilus-window.c b/src/nautilus-window.c
index 124fc57..2da42d9 100644
--- a/src/nautilus-window.c
+++ b/src/nautilus-window.c
@@ -1439,6 +1439,7 @@ nautilus_window_constructed (GObject *self)
 	GtkWidget *grid;
 	NautilusWindowSlot *slot;
 	NautilusApplication *application;
+	const gchar * desktop;
 
 	window = NAUTILUS_WINDOW (self);
 
@@ -1467,7 +1468,17 @@ nautilus_window_constructed (GObject *self)
 			 G_CALLBACK (nautilus_window_load_extension_menus), window, G_CONNECT_SWAPPED);
 
 	window->details->toolbar = create_toolbar (window);
-	gtk_window_set_titlebar (GTK_WINDOW (window), window->details->toolbar);
+	desktop = g_getenv ("XDG_CURRENT_DESKTOP");
+	if (desktop && strstr (desktop, "GNOME"))
+		gtk_window_set_titlebar (GTK_WINDOW (window), window->details->toolbar);
+	else {
+		gtk_container_add (GTK_CONTAINER (grid), window->details->toolbar);
+		gtk_widget_set_hexpand (window->details->toolbar, TRUE);
+		gtk_widget_set_vexpand (window->details->toolbar, FALSE);
+		gtk_header_bar_set_show_close_button (GTK_HEADER_BAR (window->details->toolbar),
+				 FALSE);
+		gtk_widget_show (window->details->toolbar);
+	}
 
 	window->details->content_paned = gtk_paned_new (GTK_ORIENTATION_HORIZONTAL);
 	gtk_widget_set_hexpand (window->details->content_paned, TRUE);
-- 
2.1.3

